### ISSUE TITLE
Test Suite for Differential Types

### ISSUE THREAD

**Issue by oxinabox** [JuliaDiff/ChainRulesTestUtils.jl#82]
We have a fairly solid idea of what it means to be a valid differnetials type.
Needs to support multipleation with a scalar
needs to support addition to self and to primal

It is rare to need to implement new once, but not inconceave-able.

Using ChainRulesTestUtils as a test time dependency of ChainRulesCore is not technically a circular dependency
though it is close, i think it is fine?

**Comment by willtebbutt** [comment]
Yeah, it feels fine (arguably quite natural) to make `ChainRulesTestUtils` a test-time dep for `ChainRulesCore` -- arguably all we're doing with CRTU is keeping a `TestUtils` module outside of `ChainRulesCore` to avoid complicating it.

**Comment by oxinabox** [comment]
Another thing we could do if we structure this right is check in rule_test that what ever is return is a differential type.
We kind of do that implictly

**Comment by willtebbutt** [comment]
Indeed. We could also test that it's an appropriate differential type for the primal in question (we've nearly got that functionality from `rand_tangent`).
